### PR TITLE
Add Libra post

### PR DESCRIPTION
**Pull request by zainab-ali**
Adds a blog post on Libra, similar to the talk given at TLS Copenhagen

### CODE REVIEW

File: posts/2017-06-13-libra.md [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+```scala
+val rocketMass = Quantity[FieldType[Kilogram, 1] :: HNil](40.0)
+val fuelMass = Quantity[FieldType[Kilogram, 1] :: HNil](10000.0)
+val exhaust = Quantity[FieldType[Metre, 1] :: FieldType[Second, -1] :: HNil](1000000.0)
+val distance = Quantity[FieldType[LightYear, 1] :: HNil](77.0)
+val rocketSpeed = exhaust times log(((rocketMass add fuelMass) divide rocketMass).value)
+val time: Quantity[FieldType[Year, 1] :: HNil] = distance divide rocketSpeed
+// error: type mismatch; found: FieldType[LightYear, 1] :: FieldType[Metre, -1] :: FieldType[Second, 1] :: HNil; required: FieldType[Year, 1] :: HNil
+```
+Yay! We've solved the problem!  It looks a lot more verbose than what we started with, but we can tidy this up by using the pimp my library pattern:
```
Comment (SethTisue) [inline posts/2017-06-13-libra.md]: suggest avoiding "pimp" in favor of "enrich" or "extension method"

Comment (tpolecat) [inline posts/2017-06-13-libra.md]: [bedazzler](https://www.youtube.com/watch?v=H-L-8oQ2pEA) pattern

File: posts/2017-06-13-libra.md [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+
+Ignoring the laws can result in serious problems. 
+Take the Mars Climate Orbiter, a $200 million space probe which successfully reached Mars after a year long voyage, but suddenly crashed into the Martian atmosphere on arrival.  Most components on the orbiter were using metric units, however a single component was sending instructions in Imperial units.  The other components did not detect this, and instead began a sudden descent causing the orbiter to burn up.  This was a simple unit conversion error!  It was a basic mistake that could have been easily avoided.  It should have been picked up during testing, or in the runtime validation layer.  
+
+In fact, it could even have been caught at compile time.
+
+### Compile time dimensional analysis
+
+We're going to use a similar problem to demonstrate compile time dimensional analysis.
+To fit with the theme of rocket physics, we will to tackle a rocket launch towards the distant constellation of Libra.
```
Comment (larsrh) [inline posts/2017-06-13-libra.md]: "will to"

File: posts/2017-06-13-libra.md [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+implicit val lyDivideC: Divide.Aux[LightYear, C, Year] = 
+  new Divide[LightYear, C] { type Out = Year }
+	
+implicit val lyDivideMps: Divide.Aux[LightYear, MetresPerSecond] = 
+  new Divide[LightYear, MetresPerSecond] { type Out = LightYearSecondsPerMetre }
+...
+```
+
+Unfortunately, there are an infinite number of combinations, so there are an infinite number of instances.
+Nevertheless, let's plough on with the one's we've wrote.  We can modify our rocket equation to use `add` and `divide`:
```
Comment (larsrh) [inline posts/2017-06-13-libra.md]: one's → ones

File: posts/2017-06-13-libra.md [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+
+This is good for multiples of dimensions, such as `LightYearSeconds`, but doesn't represent combinations created from division, such as `MetresPerSecond`.
+To do this, we need some way of representing integer exponents as types. We can represent integers as types using Singleton types.  We actually need these singleton types in type position.  This is supported by a new feature present in [Typelevel Scala](http://typelevel.org/scala/), called [literal types](http://docs.scala-lang.org/sips/pending/42.type.html):
+
+```scala
+scalaOrganization := "org.typelevel"
+scalacOptions += "-Yliteral-types"
+```
+
+We need to represent a key value pair of dimension and integer exponent. We could use a `Tuple` for this, but will use a shapeless `FieldType` instead.  This is similar to a `Tuple`, but is more compatable with some of shapeless's typeclasses.
```
Comment (larsrh) [inline posts/2017-06-13-libra.md]: compatable → compatible

File: posts/2017-06-13-libra.md [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+
+When we multiply and divide dimensions, we want to add or subtract from these exponents.
+
+We can use a library called [singleton ops](<https://github.com/fthomas/singleton-ops>) to do this.  This provides us with type level integer operations using the `OpInt` typeclass:
+
+```scala
+OpInt.Aux[1 + 2, 3]
+OpInt.Aux[3 * 2, 6]
+```
+It also provides a conventient alias for integer singleton types
```
Comment (larsrh) [inline posts/2017-06-13-libra.md]: conventient → convenient

**Comment by larsrh** [comment]
Merged, thanks!

**Pull request merged**